### DISCUSSION TITLE
Multiple Handles (8 handles) with proximity Connect

### DISCUSSION THREAD

**Discussion by ghost** [xyflow/xyflow#2984] [Q&A] [answered]
Can we use proximity connect with multiple handles. I have nodes with 8 handles. 4 source and 4 target. I have tried the example posted in official site. but it is not working as expected. It is not giving source handle and target handle and it is only connect 1 source and 1 target handle  right and left side respectively.
Please guide me about it.

**Comment by bcakmakoglu** [comment]
Did you add a unique id to each of your handles?
When using multiple source/target handles, each handle needs an id

**Reply by ghost** [reply]
yes without proximity connect. everything is set and working ie. each source and target handles have different Id set.

**Reply by bcakmakoglu** [reply]
What exactly isn't working then?
Is the proximity triggered to a wrong handle?
Is the handle returned in `onConnect` not the right one?

A sandbox would also help very much in debugging your issue 😄

**Reply by ghost** [reply]
How to get Handles in a custom node? I have nodes with 8 handles in it. how can i get handles related to a node to find nearest handle.
`[React Flow]: Couldn't create edge for target handle id: "undefined", edge id: Group-3852-Variable-43685. `

I am using this function to get closest node. It is only connecting to handle on left and right(default handles).

    const getClosestEdge = useCallback((node: Node) => {
    const { nodeInternals } = store.getState();
    const storeNodes = Array.from(nodeInternals.values());
    let nodeX = node.positionAbsolute ? node.positionAbsolute.x : 0;
    let nodeY = node.positionAbsolute ? node.positionAbsolute.y : 0;
    const closestNode = storeNodes.reduce(
      (
        res: {
          distance: number;
          node: Node | null;
        },
        n
      ) => {
        let nX = n.positionAbsolute ? n.positionAbsolute.x : 0;
        let nY = n.positionAbsolute ? n.positionAbsolute.y : 0;

        if (n.id !== node.id) {
          const dx = nX - nodeX;
          const dy = nY - nodeY;
          const d = Math.sqrt(dx * dx + dy * dy);

          if (
            d < res.distance &&
            d < ConstantsFlowchart.MIN_PROXIMITY_DISTANCE
          ) {
            res.distance = d;
            res.node = n;
          }
        }

        return res;
      },
      {
        distance: Number.MAX_VALUE,
        node: null,
      }
    );

    if (!closestNode.node) {
      return null;
    }

    let cX = closestNode.node.positionAbsolute
      ? closestNode.node.positionAbsolute.x
      : 0;
    const closeNodeIsSource = cX < nodeX;

    return {
      id: `${node.id}-${closestNode.node.id}`,
      data: { layer: activeLayer },
      source: closeNodeIsSource ? closestNode.node.id : node.id,
      target: closeNodeIsSource ? node.id : closestNode.node.id,
    };
    }, []);

**Reply by bcakmakoglu** [reply]
You mean the handle bounds or what exactly are you looking for when you say handles? :)

**Reply by ghost** [reply]
yeah handleBounds. I have resolved the issue but distance is not being calculated from boundries of node rather it is being calculated from top left corner of node. anyhow I have adjusted the min-distance based on (x,y) values

**Comment by ghost** [comment, marked as answer]
This function will give the edge with closest handle on closest node.

    const getClosestEdge = useCallback((node: Node) => {
    const { nodeInternals } = store.getState();
    const nodes = Array.from(nodeInternals.values());

    //Find Closest node from all nodes.
    const closestNode = nodes.reduce(
      (res: { distance: number; node: Node | null }, n) => {
        if (n.id !== node.id && n.type !== ConstantsFlowChartNodeTypes.GROUP) {
          const dx = n.position.x - node.position.x;
          const dy = n.position.y - node.position.y;
          const d = Math.sqrt(dx * dx + dy * dy);

          if (
            d < res.distance &&
            d <
              ConstantsFlowchart.MIN_PROXIMITY_DISTANCE +
                ((dx > 200 || dx < -130) && (dy < 100 || dy > -100) ? 90 : 0) //add 90 to min distance if node nodes are not in column arrange but in row arrangment.
          ) {
            // as distance being calculated from top-left corner of node so we have to increase min distance in case of nodes being arranged in a row.
            res.distance = d;
            res.node = n;
          }
        }

        return res;
      },
      {
        distance: Number.MAX_VALUE,
        node: null,
      }
    );

    if (!closestNode.node) {
      return null;
    }

    type HandleBound = {
      id: string;
      position: "string";
      x: number;
      y: number;
    };

    // find node from internal storage as node coming in parameter does not have handleBounds in it.
    let nodeToFind = nodes.find((n) => n.id === node.id);

    // node being dragged is considered as target node all other nodes will be considered as source node.
    const allHandleDistances = (nodeToFind as any)[
      Object.getOwnPropertySymbols(nodeToFind)[0]
    ].handleBounds.target.reduce(
      (closest: { [k: string]: string[] }, targetHandle: HandleBound) => {
        let closestHandle: { [k: string]: string[] } = {};

        // check distance of all source Handles on closest node to the current target handle of node being dragged.
        (closestNode.node as any)[
          Object.getOwnPropertySymbols(closestNode.node)[0]
        ].handleBounds.source.forEach((sourceBound: HandleBound) => {
          closestHandle = {
            ...closestHandle,
            //Find distance between two handles and put it in an object in format { [distance_between_handles] : [source_handle_id, target_handle_id] }
            [Math.abs(
              Math.sqrt(
                Math.pow(
                  closestNode.node?.position.x! +
                    sourceBound.x -
                    (nodeToFind?.position.x! + targetHandle.x),
                  2
                ) +
                  Math.pow(
                    closestNode.node?.position.y! +
                      sourceBound.y -
                      (nodeToFind?.position.y! + targetHandle.y),
                    2
                  )
              )
            ).toString()]: [sourceBound.id, targetHandle.id],
          };
        });

        return { ...closest, ...closestHandle };
      },
      {}
    );

    let keys = [...Object.keys(allHandleDistances)];

    //sort all distances and return edge with closest handle.
    keys.sort((a, b) => a.localeCompare(b, undefined, { numeric: true }));

    return {
      id: `L${activeLayer}_${closestNode.node.id},${
        allHandleDistances[keys[0]][0]
      }_${node.id},${allHandleDistances[keys[0]][1]}`,
      data: { layer: activeLayer },
      target: node.id,
      targetHandle: allHandleDistances[keys[0]][1],
      source: closestNode.node.id,
      sourceHandle: allHandleDistances[keys[0]][0],
    };
    }, []);

**Reply by dubey-prashant** [reply]
How did you come up with the value `90` to be added here:
 `((dx > 200 || dx < -130) && (dy < 100 || dy > -100) ? 90 : 0) `

**Reply by ghost** [reply]
@dubey-prashant ummm, My bad, I should have added comments for that statement. Now I forgot why I added 90 there. I guess I did that if nodes are in row arrangement and distance being calculated from top-left corner of node , the distance will always be within the width of first node in row arrangement. check the image and proximity is shown by boxes..  both nodes need to overlap to fall in proximity range.. I think thats why I did that .. I hope..
![Screenshot from 2023-10-17 11-31-49](https://github.com/wbkd/react-flow/assets/110381868/3071c407-6f19-4b1a-8e8d-be35f64813ac)
